### PR TITLE
SONARJAVA-3096 Rule S2068, S2092, S2115, S3330: fix FN on variable re…

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
…solution with single assignment

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/helpers/IdentifierUtils.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import static org.sonar.java.checks.helpers.ReassignmentFinder.getInitializerOrExpression;
+import static org.sonar.java.checks.helpers.ReassignmentFinder.getReassignments;
+
+public class IdentifierUtils {
+  private IdentifierUtils() {
+    // This class only contains static methods
+  }
+
+  public static <T> ValueResolution<T> getUniqueAssignmentValueResolution(ExpressionTree expression, Function<ExpressionTree,T> resolver) {
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/helpers/IdentifierUtils.java]: This function would work not only for identifiers, I think it should go into  `ExpressionUtils`, also this name is very long, perhaps something like `getConstantValue` would be sufficient

File: java-checks/src/main/java/org/sonar/java/checks/helpers/IdentifierUtils.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    return valueResolution(expression, resolver, new ValueResolution<>());
+  }
+
+  private static <T> ValueResolution<T> valueResolution(ExpressionTree expression, Function<ExpressionTree,T> resolver, ValueResolution<T> valueResolution) {
```
Comment (saberduck) [inline java-checks/src/main/java/org/sonar/java/checks/helpers/IdentifierUtils.java]: I think we could use `ConstantUtils.resolveAsConstant` as universal resolver and remove the need for `resolver` parameter. Users can cast the return value on the call site if needed.

**Comment by saberduck** [comment]
Usage search for `Constant.resolveAsString` shows more rules which could benefit from this mechanism like S3331, S5122, S4432, S4925, S4719 . I think especially recent rules in `checks.security` package should be updated to use this.

**Comment by andrea-guarino-sonarsource** [comment]
As discussed, I will create a separate ticket to update other rules using `Constant.resolveAsString`

**Pull request merged**